src/hardware/shared_adc.rs

```diff
@@ -0,0 +1,80 @@
+use embedded_hal::adc::{Channel, OneShot};
+use stm32h7xx_hal as hal;
+
+#[derive(Debug, Copy, Clone)]
+pub enum AdcError {
+    Allocated,
+}
+
+pub struct AdcChannel<'a, Adc, PIN> {
+    pin: PIN,
+    slope: f32,
+    adc: &'a cortex_m::interrupt::Mutex<
+        core::cell::RefCell<hal::adc::Adc<Adc, hal::adc::Enabled>>,
+    >,
+}
+
+impl<'a, Adc, PIN> AdcChannel<'a, Adc, PIN>
+where
+    PIN: Channel<Adc, ID = u8>,
+    hal::adc::Adc<Adc, hal::adc::Enabled>: OneShot<Adc, u32, PIN>,
+    <hal::adc::Adc<Adc, hal::adc::Enabled> as OneShot<Adc, u32, PIN>>::Error:
+        core::fmt::Debug,
+{
+    pub fn read(&mut self) -> f32 {
+        cortex_m::interrupt::free(|cs| {
```

Comment:
The current synchronization mechanism is a critical section, which is likely undesriable.

Currently, the only user of ADC channels is Pounder, which uses them for:
1. Reading input power measurements
2. Reading auxiliary ADC inputs

In the future, the ADC will also be used for:
1. Measuring internal CPU temperature

We need to determine a proper synchronization mechanism for this.

@jordens When are the auxiliary pounder ADC inputs supposed to be read?

Reply:
It looks like the input power measurements and the CPU temperature will likely all be measured in the same `telemetry` task, so we can likely use an atomic-check "mutex" that just double-checks that we haven't pre-empted ourselves (similar to shared-bus-rtic's concepts).

Reply:
Yes. Mutex (advantage no blocking of higher prio tasks, disadvantage possible deadlock) would be better than CS (blocking of higher prio tasks) here.
But alternatively we could just let RTIC handle it as an ADC resource. For that move it into `StabilizerDevices` (IIRC) and move all three ADCs into it and require that the ADC resource be `lock()` ed when a tasks wants to sample a (e.g. local resource) channel. API would be `ADCsResource.read(&AdcChannel)`. Since ADC1/2/3 are used all over there doesn't seem a natural need or way to split them.

Reply:
The main purpose of this design is such that the ADC1/ADC2/ADC3 peripherals do **not** need to be passed around because that's painful from an end-user perspective.

Reply:
Ack